### PR TITLE
Migrate from rest to full websocket

### LINKED ISSUE #153: Przepisać RestAPI na WebSocket

### LINKED ISSUE #186: Podpiąć stare elementy z Resta na Sockety
Jeżeli ustalimy że przechodzimy na websockety to trzeba całą apkę przebudować żeby stare funkcjonalności działały

### PR DESCRIPTION
**Pull request by rozchlastywacz**
I replaced resp api functionalities with websockets, so you can now:
- create lecture with tutor name and it is immidiately subrscribed to certain events
- create student with link and unique nick, name, surname and they are immidiately subscribed to certain events
- check if lecture link is valid
- reconnect lecture with lecture id
- reconnect student with student id
- delete lecture, which also deletes all coennected students and closes all connections
- delete student, which removes all student data and closes connection
- as lecturer get whole student list from server

Waitin' for revs'
Misato-Katsuragi-Is-Bored-Browsing-The-Internet-On-Evangelion-Anime

### CODE REVIEW

File: server/Student.ts [hunk 1]
```diff
@@ -90,6 +93,24 @@ class Student extends EventEmitter {
...
+        }
+        this.handleRodo();
+        this.lecture.studentList.deleteStudent(this.id);
+    }
+
+    handleRodo(){
+        this.name = "";
+        this.surname = "";
+        this.nick = "";
+    }
```
Comment (ostatni5) [inline server/Student.ts:112]: What purpose is this function if a student gets deleted?

File: server/Student.ts [hunk 2]
```diff
@@ -43,14 +43,17 @@ class Student extends EventEmitter {
                 case "send_reaction":
                     this.handlerSendReaction(parsed);
                     break;
+                case "kill_me":
```
Comment (ostatni5) [inline server/Student.ts:46]: Maybe would be better  `delete`?

File: server/StudentList.ts [hunk 3]
```diff
@@ -40,5 +41,12 @@ class StudentList extends EventEmitter {
         return (student1.name === student2.name && student1.surname === student2.surname && student1.nick === student2.nick)
     }
 
+    includesStudent(student: Student): boolean{
```
Comment (ostatni5) [inline server/StudentList.ts:44]: It is StudentList it only should include Students.
I propose name: `includes`

File: server/websockets.ts [hunk 4]
```diff
@@ -29,38 +43,104 @@ const setupWebSocketServer = () => {
...
 };
 
-function handlerSubscribeStudent(parsed: StudentSubPayload, ws: WebSocketClient) {
-    const selectedLecture: Lecture | undefined = linkLectureMap.get(parsed.data.lecture_link);
-    const selectedStudent: Student | undefined = selectedLecture?.studentList.getStudent(parsed.data.student_id);
+function handlerCreateStudent(parsed: StudentCreateRequestPayload, ws: WebSocketClient): boolean {
+    if (linkLectureMap.has(parsed.data.lecture_link)) {
+        const selectedLecture: Lecture = linkLectureMap.get(parsed.data.lecture_link);
+        const student: Student = new Student(parsed.data.nick, parsed.data.name, parsed.data.surname, selectedLecture);
+        if(!selectedLecture.studentList.includesStudent(student)){
```
Comment (ostatni5) [inline server/websockets.ts:50]:
```suggestion
        const student: Student = new Student(parsed.data.nick, parsed.data.name, parsed.data.surname, selectedLecture);
        
        if(!selectedLecture.studentList.includesStudent(student)){

```

File: server/websockets.ts [hunk 5]
```diff
@@ -29,38 +43,104 @@ const setupWebSocketServer = () => {
...
+                event: "student_created",
+                data: {
+                    student_id: student.id
+                }
+            };
+            ws.send(JSON.stringify(payload));
+            return true;
+        }
+    }
+    const payload: Payload = {
```
Comment (ostatni5) [inline server/websockets.ts:63]:
```suggestion
    }
    
    const payload: Payload = {

```

**Review by ostatni5 — APPROVED** [review]
Good job.
Consider my comments.

obraz

**Review by ostatni5 — CHANGES_REQUESTED** [review]
Squash commits before the merge.

Also would be great u created a type that has possible events.
ex:
`type Easing = "ease-in" | "ease-out" | "ease-in-out";`

File: server/@types/payloads/types.d.ts [hunk 6]
```diff
@@ -3,19 +3,32 @@ declare module 'payloads' {
         event: string
     }
 
-    interface LectureSubPayload extends Payload {
+    interface LectureCreateRequestPayload extends Payload {
         data: {
-            lecture_id: string
+            lecturer: string
```
[the remark this answers is not in the document]
Comment (rozchlastywacz) [reply server/@types/payloads/types.d.ts:8]: coz lecture id is not required with websockets and tutor field in lecture wasn`t used until now

File: server/Lecture.ts [hunk 7]
```diff
@@ -125,6 +147,49 @@ class Lecture {
...
+            this.wsc?.close(1000, "Lecturer requested shutdown");
+        }
+        this.studentList.asArray().forEach((student: Student) => {
+            student.handleRodo();
+            student.wsc?.send(JSON.stringify(payload));
+            if(!student.wsc?.isClosed){
+                student.wsc?.close(1000, "Lecturer requested shutdown");
+            }
+        });
+        // Rodo
```
[the remark this answers is not in the document]
Comment (rozchlastywacz) [reply server/Lecture.ts:165]: u need to be sure that data inserted by user is deleted when he wants to

File: server/Student.ts (hunk 1 above)
Comment (rozchlastywacz) [reply server/Student.ts:112]: to be sure that personal data is deleted, so I overwrite it

File: server/Student.ts (hunk 2 above)
Comment (rozchlastywacz) [reply server/Student.ts:46]: image

File: server/StudentList.ts (hunk 3 above)
Comment (rozchlastywacz) [reply server/StudentList.ts:44]: u r ->

File: server/websockets.ts (hunk 4 above)
Comment (rozchlastywacz) [reply server/websockets.ts:50]: those are white spaces only?

File: server/websockets.ts (hunk 5 above)
Comment (rozchlastywacz) [reply server/websockets.ts:63]: also white spaces?

File: server/websockets.ts (hunk 4 above)
Comment (ostatni5) [reply server/websockets.ts:50]: just enter

File: server/@types/payloads/types.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ostatni5) [reply server/@types/payloads/types.d.ts:8]: why not name it `lecturer_id`, to keep convention?

File: server/Lecture.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ostatni5) [reply server/Lecture.ts:165]: meh I don't think so this comment is necessary

File: server/Student.ts (hunk 1 above)
Comment (ostatni5) [reply server/Student.ts:112]: GDPR not RODO
https://en.wikipedia.org/wiki/General_Data_Protection_Regulation

And i dont think we need such function we don't store users in database.

File: server/websockets.ts (hunk 5 above)
Comment (ostatni5) [reply server/websockets.ts:63]: just enter

File: server/@types/payloads/types.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (rozchlastywacz) [reply server/@types/payloads/types.d.ts:8]: coz id and lecturer are different types of strings
image

File: server/StudentList.ts (hunk 3 above)
Comment (rozchlastywacz) [reply server/StudentList.ts:44]: you are right xD

File: server/@types/payloads/types.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ostatni5) [reply server/@types/payloads/types.d.ts:8]: So what represents `lecturer `, because that is not clear to me.

**Review by ostatni5 — APPROVED** [review]
a

**Review by Derstilon — CHANGES_REQUESTED** [review]
We need to connect all things that used Rest to Sockets

**Review by Derstilon — COMMENTED** [review]
Should a student disappear from the list after his browser card is closed?

**Review by Derstilon — APPROVED** [review]
image

**Pull request merged**